cli/swc_util.rs

```diff
@@ -0,0 +1,112 @@
+use crate::file_fetcher::SourceFile;
+use crate::msg::MediaType;
+use std::sync::Arc;
+use std::sync::RwLock;
+use swc_common;
+use swc_common::errors::Diagnostic;
+use swc_common::errors::DiagnosticBuilder;
+use swc_common::errors::Emitter;
+use swc_common::errors::Handler;
+use swc_common::errors::HandlerFlags;
+use swc_common::FileName;
+use swc_common::SourceMap;
+use swc_ecma_parser::lexer::Lexer;
+use swc_ecma_parser::JscTarget;
+use swc_ecma_parser::Parser;
+use swc_ecma_parser::Session;
+use swc_ecma_parser::SourceFileInput;
+use swc_ecma_parser::Syntax;
+use swc_ecma_parser::TsConfig;
+
+pub type SwcDiagnostics = Vec<Diagnostic>;
+
+#[derive(Clone, Default)]
+pub(crate) struct BufferedError(Arc<RwLock<SwcDiagnostics>>);
+
+impl Emitter for BufferedError {
+  fn emit(&mut self, db: &DiagnosticBuilder) {
+    self.0.write().unwrap().push((**db).clone());
+  }
+}
+
+impl From<BufferedError> for Vec<Diagnostic> {
+  fn from(buf: BufferedError) -> Self {
+    let s = buf.0.read().unwrap();
+    s.clone()
+  }
+}
+
+pub struct Compiler {
+  buffered_error: BufferedError,
+  pub source_map: Arc<SourceMap>,
+  pub handler: Handler,
+}
+
+impl Compiler {
+  pub fn new() -> Self {
+    let buffered_error = BufferedError::default();
+
+    let handler = Handler::with_emitter_and_flags(
+      Box::new(buffered_error.clone()),
+      HandlerFlags {
+        dont_buffer_diagnostics: true,
+        can_emit_warnings: true,
+        ..Default::default()
+      },
+    );
+
+    Compiler {
+      buffered_error,
+      source_map: Arc::new(SourceMap::default()),
+      handler,
+    }
+  }
+
+  pub fn json_ast(
+    &mut self,
+    source_file: SourceFile,
+  ) -> Result<String, SwcDiagnostics> {
+    swc_common::GLOBALS.set(&swc_common::Globals::new(), || {
+      let swc_source_file = self.source_map.new_source_file(
+        FileName::Custom(source_file.url.as_str().into()),
+        std::str::from_utf8(&source_file.source_code)
+          .unwrap()
+          .into(),
+      );
+
+      let buffered_err = self.buffered_error.clone();
+      let session = Session {
+        handler: &self.handler,
+      };
+
+      let syntax = match source_file.media_type {
+        MediaType::TypeScript => {
+          let mut ts_config = TsConfig::default();
+          ts_config.dynamic_import = true;
+          Syntax::Typescript(ts_config)
+        }
+        _ => Syntax::Es(Default::default()),
+      };
+
+      let lexer = Lexer::new(
+        session,
+        syntax,
+        JscTarget::Es2019,
+        SourceFileInput::from(&*swc_source_file),
+        None,
+      );
```

Comment:
The last argument is "comments"; instead of `None` we can pass `Some(Comments::default())` however, comments do not show in AST anyway. I wonder if there's some helper in `swc` that can annotate AST with comments or do we have to do that manually. CC @dsherret for opinion

Reply:
https://github.com/swc-project/swc/issues/506

Reply:
Yeah, the comments are stored separately from the AST in those maps for start pos -> leading and end pos -> trailing.

What's the use case for this command? Are comments necessary?

I'm not sure how easy this would be to do in rust, but you could traverse the json and check the `start` and `end` properties to fill in a custom `leadingComments` and `trailingComments` property similar to what Babel does:

![image](https://user-images.githubusercontent.com/1609021/76864660-d8eb2300-6837-11ea-9c23-d4d276360258.png)

The basic rule I've found for handling nodes with the same start or end is that nodes higher up in the tree get the comment placed on them. Example `/*c*/ someIdentifier;` would have an expression statement and identifier at the same start, but only the expression statement would get the comment. See [here](https://astexplorer.net/#/gist/64961f4b151c2c9162af97d94a09947d/281ccd9466ffd892900ebf41cc35d8ce8e424a26).

Reply:
> What's the use case for this command? Are comments necessary?

We're thinking about implementing `deno doc` first - a command which would print out JSdocs for all exported members of the module.

> The basic rule I've found for handling nodes with the same start or end is that nodes higher up in the tree get the comment placed on them. Example `/*c*/ someIdentifier;` would have an expression statement and identifier at the same start, but only the expression statement would get the comment. See [here](https://astexplorer.net/#/gist/64961f4b151c2c9162af97d94a09947d/281ccd9466ffd892900ebf41cc35d8ce8e424a26).

Thanks, that's helpful

Reply:
Ah, I see. An overview of jsdocs in the other compilers in case you guys aren't familiar in order to get some ideas:

* Babel has jsdocs in that `"leadingComments"` property and you need to manually have a way to tell if a comment is a js doc.
* swc stores has it in the leading comment map. Also need to manually develop a way to tell if a comment is a js doc.
* TypeScript stores them in an internal property called `jsDoc`.

    ![image](https://user-images.githubusercontent.com/1609021/76867006-4b113700-683b-11ea-8a15-8ff3dd86c384.png)

Reply:
@dsherret, thanks again, that's also very helpful! 

> * swc stores has it in the leading comment map. Also need to manually develop a way to tell if a comment is a js doc.

Just hit this case, it looks like `swc` strips leading and trailing comment tokens.

Example:
```ts
/** Hello there, this is a multiline JSdoc.
 * 
 * It has many lines
 * 
 * Or not that many?
 */
export function foo(a: string, b: number): void {
    console.log("Hello world");
}

/** This is single line JSdoc */
export function bar(a: string, b: number): void {
    console.log("Hello world");
}
```

And here's output from swc's `leading_comments`
```rust
[
    Comment {
        kind: Block,
        span: Span {
            lo: BytePos(
                39,
            ),
            hi: BytePos(
                136,
            ),
            ctxt: #0,
        },
        text: "* Hello there, this is a multiline JSdoc.\n * \n * It has many lines\n * \n * Or not that many?\n ",
    },
]
[
    Comment {
        kind: Block,
        span: Span {
            lo: BytePos(
                222,
            ),
            hi: BytePos(
                254,
            ),
            ctxt: #0,
        },
        text: "* This is single line JSdoc ",
    },
]
```

Reply:
Yeah, all the compilers do that. Babel example:

![image](https://user-images.githubusercontent.com/1609021/76868486-67ae6e80-683d-11ea-9e26-18aede54c28c.png)


In swc, the `/*` and `*/` is implied by the `kind` property being `CommentKind::Block`. For `CommentKind::Line` I believe it doesn't include the leading `//` either.

Reply:
According to [`JSDoc website`](https://jsdoc.app/about-getting-started.html) all JSDoc comments should start with `/**` so I guess veryfing that parsed comment starts with `*` is good enough.

Reply:
Yup, I believe that's correct 👍